### PR TITLE
feat: --psql flag for providing psql connection to db

### PR DESCRIPTION
**Pull request by bayandin**
This PR adds `--psql` flag to the following commands:
- `projects create` 
- `branches create` 
- `connection-string`

The flag makes neonctl launch psql session for a given project/branch/connection string.
For non-idempotent operations (`projects create` and `branches create`), it happens after the standard output (with new project/branch information).
For idempotent `connection-string`, it launches psql instead of the output. 

Additional psql arguments can be passed after `--`:

```bash
neonctl branches create --project-id orange-field-03303460 --psql -- -c "SHOW neon.timeline_id"  
branch
┌────────────────────────┬────────────────────────┬─────────┬──────────────────────┬──────────────────────┐
│ Id                     │ Name                   │ Primary │ Created At           │ Updated At           │
├────────────────────────┼────────────────────────┼─────────┼──────────────────────┼──────────────────────┤
│ br-dark-water-38911821 │ br-dark-water-38911821 │ false   │ 2023-08-14T09:50:41Z │ 2023-08-14T09:50:41Z │
└────────────────────────┴────────────────────────┴─────────┴──────────────────────┴──────────────────────┘
endpoints
┌────────────────────────────┬──────────────────────┐
│ Id                         │ Created At           │
├────────────────────────────┼──────────────────────┤
│ ep-soft-waterfall-20122714 │ 2023-08-14T09:50:41Z │
└────────────────────────────┴──────────────────────┘
connection_uris
┌────────────────────────────────────────────────────────────────────────────────────────────┐
│ Connection Uri                                                                             │
├────────────────────────────────────────────────────────────────────────────────────────────┤
│ postgres://bayandin:<redacted>@ep-soft-waterfall-20122714.us-east-2.aws.neon.tech/neondb │
└────────────────────────────────────────────────────────────────────────────────────────────┘
INFO: Connecting to the database using psql...
         neon.timeline_id         
----------------------------------
 89a6ed1fb7c8e6e27352b5ecbaa0dbeb
(1 row)
```

### CODE REVIEW

File: src/utils/psql.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { log } from '../log.js';
+
+export const psql = async (
+  connection_uri: string,
+) => {
+  const { execSync, spawnSync } = await import('child_process');
+
+  const command = process.platform === 'win32' ? 'where' : 'which';
```
Comment (bayandin) [inline src/utils/psql.ts:8]: I haven't tested this on Windows

File: src/utils/psql.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { log } from '../log.js';
+
+export const psql = async (
+  connection_uri: string,
+) => {
+  const { execSync, spawnSync } = await import('child_process');
```
Comment (bayandin) [inline src/utils/psql.ts:6]: I'm not sure if it's ok to use a synchronous version of `exec` and `spaws`, but it works.

File: src/utils/psql.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    stdio: 'inherit',
+  });
+};
+
+export const psqlArgs = (
+  argv: string[]
+) => {
+    const doubleDashIndex = argv.indexOf('--');
+    return doubleDashIndex === -1 ? [] : argv.slice(doubleDashIndex + 1);
+}
```
Comment (bayandin) [inline src/utils/psql.ts:27]: I haven't found a way how to do it with `yargs`, so I'm parsing raw `process.argv`

**Comment by duskpoet** [comment]
What would be the benefit of this approach over using a command substitution?
Like `psql $(neonctl cs)`

**Comment by bayandin** [comment]
> What would be the benefit of this approach over using a command substitution? Like `psql $(neonctl cs)`

From the usability point of view, you won't need to switch context if you're already working with `neonctl`. If you're typing it, it's easier to add `... --psql` than `psql $(...)`.  For one-off commands (especially in scripts), there's no difference.

Initially, I added it to match `neonctl projects create --psql` behaviour (when I ran out of projects limit), but I still find it quite useful.

I also would like to add `--psql` flag to `neon branches create`, but first would like to get some feedback about it.

**Comment by bayandin** [comment]
I've seen Lacho asking around about `neonctl` UX. 
What do you think about `--psql` flag @lpetkov?

**Comment by duskpoet** [comment]
I know some people prefer pgcli over psql, some prefer GUI clients, so I feel like this could be a rabbit-hole

File: src/utils/psql.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import { log } from '../log.js';
+
+export const psql = async (
+  connection_uri: string,
+  args: string[] = [],
+) => {
+  const { execSync, spawn } = await import('child_process');
```
Comment (duskpoet) [inline src/utils/psql.ts:7]: no need for async import here, let's hoist to the top level

File: src/utils/psql.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+  const which = process.platform === 'win32' ? 'where' : 'which';
```
Comment (duskpoet) [inline src/utils/psql.ts:9]: let's use this package https://www.npmjs.com/package/which

**Comment by duskpoet** [comment]
I am wondering, how can a test for that flag look...

File: src/utils/psql.ts (hunk 3 above)
Comment (bayandin) [reply src/utils/psql.ts]: Thanks for the suggestion!

**Comment by bayandin** [comment]
> I am wondering, how can a test for that flag look...

Added tests with mocked `psql`

File: src/commands/branches.test.ts [hunk 6]
```diff
@@ -26,6 +26,47 @@ describe('branches', () => {
...
+      'branches',
+      'create',
+      '--project-id',
+      'test',
+      '--name',
+      'test_branch',
+      '--psql',
+    ],
+    env: {
+      PATH: `mocks/bin:${process.env.PATH}`,
```
Comment (duskpoet) [inline src/commands/branches.test.ts:41]: I think it's ok to add this to `testCliCommand` code, it's pretty convenient way to mock external tools! And we most likely would never need to switch between mocked and real binary

**Review by duskpoet — APPROVED** [review]
Cool stuff, thanks for contributing! I like the idea of mocking binaries, pretty flexible!

**Comment by duskpoet** [comment]
One more thing: make sure to format merge to main commit with conventional commit, so it would generate proper changelog and next release version. Would be something like `feat: --psql flag for project creation and connection string`

**Comment by lpetkov** [comment]
I'm late to the party. LGTM, just one thought: if we want to scale this in the future, doesn't it make sense for the flag to be called `--client=psql`? And then over time add `--client=dbeaver` etc

**Comment by duskpoet** [comment]
Probably @stepashka can chime in too

**Comment by duskpoet** [comment]
> I'm late to the party. LGTM, just one thought: if we want to scale this in the future, doesn't it make sense for the flag to be called `--client=psql`? And then over time add `--client=dbeaver` etc

honestly, I would like to avoid going that path, cause it's too much work for singular cases, and not really clear for what purpose, cause you can just substitute to whatever you prefer to use, like 
```
psql $(neonctl cs)
pgcli $(neonctl cs)
...
```

**Review by duskpoet — APPROVED** [review]

**Comment by duskpoet** [comment]
I have figured out one more reason to use this approach instead of shell substitution: when you use shell substitution it suppresses stderr output, so you won't see any logs or error messages, this way it all works good

**Pull request merged**